audioipc/src/shm.rs

```diff
@@ -7,22 +7,96 @@ use crate::errors::*;
 use memmap::{Mmap, MmapMut, MmapOptions};
 use std::cell::UnsafeCell;
 use std::fs::{remove_file, File, OpenOptions};
-use std::path::Path;
+use std::path::PathBuf;
 use std::sync::{atomic, Arc};
+use std::convert::TryInto;
+use std::env::temp_dir;
+
+fn open_shm_file(id: &str) -> Result<File> {
+    if cfg!(target_os = "linux") {
+        let mut path = PathBuf::from("/dev/shm");
+        path.push(id);
+
+        if let Ok(file) = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create_new(true)
+            .open(&path) {
+                let _ = remove_file(&path);
+                return Ok(file);
+            }
+    };
+
+    let mut path = temp_dir();
+    path.push(id);
+
+    let file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create_new(true)
+        .open(&path)?;
+
+    let _ = remove_file(&path);
+    Ok(file)
+}
+
+#[cfg(unix)]
+fn allocate_file(file: &File, size: usize) -> Result<()> {
+    use std::os::unix::io::AsRawFd;
+
+    // Try Linux-specific fallocate.
+    #[cfg(target_os = "linux")]
+    unsafe {
+        if libc::fallocate(file.as_raw_fd(), 0, 0, size.try_into().unwrap()) == 0 {
+            return Ok(())
+        }
+    }
+
+    // Try macOS-specific fcntl.
+    #[cfg(target_os = "macos")]
+    unsafe {
+        let params = libc::fstore_t {
+            fst_flags: libc::F_ALLOCATEALL,
+            fst_posmode: libc::F_PEOFPOSMODE,
+            fst_offset: 0,
+            fst_length: size.try_into().unwrap(),
+            fst_bytesalloc: 0,
+        };
+        let r = libc::fcntl(file.as_raw_fd(), libc::F_PREALLOCATE, &params);
+        if r == 0 {
+            // F_PREALLOCATE doesn't update the file's size, so set now.
+            file.set_len(size.try_into().unwrap())?;
```

Comment:
No fullback plan on Mac if it fails?

Reply:
We still fall back to the set_len at the bottom of open_shm_file, which was our original approach before adding these refinements.

Reply:
I think there's a slight flaw in my approach here, though, which is that we'll always fallback if any of these preallocation attempts fail due any error including low disk space.  For the low disk space case, we really want to hard error rather than falling back, since we're explicitly trying to avoid creating sparse files that later cause our shmem to fault with SIGBUS due to the lazy allocation failing later.

I'll update my patch and request review again.  Thanks!